plugin/static_account.go

```diff
@@ -0,0 +1,344 @@
+		if a.TokenGen == nil {
+			return nil, fmt.Errorf("unexpected invalid access_token static account has no TokenGen")
+		}
+		if !strutil.EquivalentSlices(updateInput.scopes, a.TokenGen.Scopes) {
+			b.Logger().Debug("detected scopes change, updating scopes for static account")
+			a.TokenGen.Scopes = updateInput.scopes
+			madeChange = true
+		}
+	}
+
+	if !madeChange {
+		return []string{"no changes to bindings or token_scopes detected, no update needed"}, nil
+	}
+
+	if err := a.save(ctx, req.Storage); err != nil {
+		return nil, err
+	}
+
+	b.tryDeleteWALs(ctx, req.Storage, walIds...)
+	return
+
+}
+
+func (b *backend) updateBindingsForStaticAccount(ctx context.Context, req *logical.Request, a *StaticAccount, newBindings ResourceBindings) ([]string, error) {
+	oldBindings := a.Bindings
+	bindsToAdd := newBindings.sub(oldBindings)
+	bindsToRemove := oldBindings.sub(newBindings)
+
+	b.Logger().Debug("updating bindings for static account")
+	walIds, err := b.addWalsForStaticAccountBindings(ctx, req, a, bindsToAdd, bindsToRemove)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := b.createIamBindings(ctx, req, a.EmailOrId, bindsToAdd); err != nil {
+		return nil, err
+	}
+
+	if err := b.removeBindings(ctx, req, a.EmailOrId, bindsToRemove); err != nil {
+		return nil, err
+	}
+
+	return walIds, nil
+}
+
+// addWalsForStaticAccountResources creates WALs to clean up a roleset's service account, bindings, and a key if needed.
+func (b *backend) addWalsForStaticAccountResources(ctx context.Context, req *logical.Request, staticAcctName string, boundResources *gcpAccountResources) (walIds []string, err error) {
+	if boundResources == nil {
+		b.Logger().Debug("skip WALs for nil GCP account resources")
+		return nil, nil
+	}
+
+	walIds = make([]string, 0, len(boundResources.bindings)+1)
+	for resName, roles := range boundResources.bindings {
+		walId, err := framework.PutWAL(ctx, req.Storage, walTypeIamPolicyDiff, &walIamPolicyStaticAccount{
+			StaticAccount: staticAcctName,
+			AccountId:     boundResources.accountId,
+			Resource:      resName,
+			RolesAdded:    roles.ToSlice(),
+		})
+		if err != nil {
+			return walIds, errwrap.Wrapf("unable to create WAL entry to clean up service account bindings: {{err}}", err)
+		}
+		walIds = append(walIds, walId)
+	}
+
+	if boundResources.tokenGen != nil {
+		walId, err := b.addWalStaticAccountServiceAccountKey(ctx, req, staticAcctName, &boundResources.accountId, boundResources.tokenGen.KeyName)
+		if err != nil {
+			return walIds, err
+		}
+		walIds = append(walIds, walId)
+	}
+	return walIds, nil
+}
+
+func (b *backend) addWalsForStaticAccountBindings(ctx context.Context, req *logical.Request, a *StaticAccount, removed, added ResourceBindings) (walIds []string, err error) {
+	walIds = make([]string, 0, len(removed)+len(added))
+
+	// Add WALs for resources in added bindings
+	for resource, rolesAdded := range added {
+		walEntry := &walIamPolicyStaticAccount{
+			AccountId:  a.ServiceAccountId,
```

Comment:
I think we'll want to add `StaticAccount: a.Name` to this WAL entry. If not, there is an eventual panic in `serviceAccountPolicyDiffRollback()` (rollback.go#L263). I observed this when updating bindings on a static account to a role that caused an error.